### ISSUE TITLE
What is this about?

### ISSUE THREAD

**Issue by matze** [ufo-kit/tofuweb#1]
I took the time in the train to quickly sketch this basic reconstruction webservice. It is not meant to be productive in any way nor will it ever be guaranteed to become that. However, I see this toy as a playground for further discussions about how we handle metadata and what workflows we must support. So @tfarago and @Threadmonkey, I am eager to hear comments and suggestions.

Some points I noticed while developing this:
- I have absolutely no idea about web frontend development, so everything there might be total crap.
- `tofu.tomo.reco` enforces a certain style of output files and at least in this case here is too restricting. For the video I patched `tofu` so that I could return a single multi TIFF. The relevant issue is covered [here](https://github.com/ufo-kit/tofu/issues/31).
- Using `multiprocessing` is very likely not the way to go to trigger reconstructions. I suppose something based on Redis or Celery is more appropriate.
- Parameterization of the reconstruction is missing but until we are absolutely sure what constitutes required parameters, I will leave this as is.

---
### Reconstruction meta data
- [x] Axis position/rotation
- [ ] Reconstruction method (currently FBP, DFI and SART)
#### Provenance-related
- [x] Reconstruction software + version
- [x] Reconstruction time
- [ ] System specification (?)

**Comment by Threadmonkey** [comment]
Point 1
Awesome. All this in one train ride? OK lets get started. This is a very nice playground. Web front end development is also not my kung fu, such that I cannot really judge. (It looks very nice)

Point 2
I should learn to use tofu, such that I currently also suck in this area.

Point 3
Now, multiprocessing is fully in my interest zone. Celery seems to have a high level, such that no low level programming is necessary. I suggest to batch the jobs locally, maybe with a local celery worker to maybe later move to distributed workers. By the way: how much time does a reconstruction take on a CPU and on a GPU in average for a normal 3D tomogram?

Point 4
From my bloody noob user point of view, there is a list of measurements (Digital Objects) and each entry has a reconstruct button. How complicated can it be? hmm I learned, that the axis is important, such that I maybe want to set that parameter.
Actually who is reconstructing things for users? Is it always an experienced scientist or is this a job for the intern?

Some metadata I have in mind are the reconstruction type and the reconstruction software (+version).

I think that reconstruction run time and system specification is nice to have for comparison and performance analysis. For the web it would not be necessary. Nice to haves could stay maybe stay nice-to-have.

After the ASTOR Working day I make my hands dirty and try to reconstruct a volume.

**Comment by matze** [comment]
> Now, multiprocessing is fully in my interest zone. Celery seems to have a high level, such that no low level programming is necessary. I suggest to batch the jobs locally, maybe with a local celery worker to maybe later move to distributed workers.

Indeed. I was just pointing out the major flaw that there is right now. But on the other hand it is actually working and won't crash the server instance itself. But you are right, moving to a distributed task queue allows us to scale much easier.

> By the way: how much time does a reconstruction take on a CPU and on a GPU in average for a normal 3D tomogram?

I give you ballpark numbers here: hours for CPU (the only possibility on my laptop, so I cheated in the demo by reconstructing from only two projections), minutes for GPU.

> Actually who is reconstructing things for users? Is it always an experienced scientist or is this a job for the intern?

I'd argue that this should be possible for everyone. For simple setups as we have, the user should be able to reconstruct himself.

> Some metadata I have in mind are the reconstruction type and the reconstruction software (+version). I think that reconstruction run time and system specification is nice to have for comparison and performance analysis. For the web it would not be necessary. Nice to haves could stay maybe stay nice-to-have.

That's a good idea, I will update the issue description.

**Comment by tfarago** [comment]
I liked the video and will play with this soon.

**Comment by tfarago** [comment]
1. Likewise, but from what I have seen it is more than enough for now
2. Yep
3. Sure
4. The parameters needed for the reconstruction, nothing else I would say, so e.g. for FBP it's just one - rotation axis. I am currently also leaning towards using `tofu` on a script-per-step basis, so no one-kilometer-long commands via `ufo-reconstruct` with flat correction, sino generation and everything... We need to think about something smarter for connecting the logical steps anyway otherwise it will blow up completely also there.
